File: rust/herb-linter/src/rules/html_no_positive_tab_index.rs
use crate::utils::tag_utils::{get_attribute, get_static_attribute_value};

use herb::nodes::HTMLOpenTagNode;
use herb::Visitor;

rule_visitor!(NoPositiveTabIndexVisitor);
define_parser_rule!(HTMLNoPositiveTabIndexRule, "html-no-positive-tab-index", Error, NoPositiveTabIndexVisitor);

impl Visitor for NoPositiveTabIndexVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    if let Some(tabindex_attribute) = get_attribute(node, "tabindex") {
      if let Some(value) = get_static_attribute_value(tabindex_attribute) {
        if let Ok(tab_index) = value.trim().parse::<i32>() {
          if tab_index > 0 {
            self.add_offense(
              "Do not use positive `tabindex` values as they are error prone and can severely disrupt navigation experience for keyboard users. Use `tabindex=\"0\"` to make an element focusable or `tabindex=\"-1\"` to remove it from the tab sequence.".to_string(),
              tabindex_attribute.location.clone(),
            );
          }
        }
      }
    }

    self.walk_html_open_tag_node(node);
  }
}

File: javascript/packages/linter/src/rules/html-no-positive-tab-index.ts
import { ParserRule } from "../types.js"
import { AttributeVisitorMixin, StaticAttributeStaticValueParams } from "./rule-utils.js"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ParseResult } from "@herb-tools/core"

class NoPositiveTabIndexVisitor extends AttributeVisitorMixin {
  protected checkStaticAttributeStaticValue({ attributeName, attributeValue, attributeNode }: StaticAttributeStaticValueParams): void {
    if (attributeName !== "tabindex") return

    const tabIndexValue = parseInt(attributeValue, 10)

    if (!isNaN(tabIndexValue) && tabIndexValue > 0) {
      this.addOffense(
        `Do not use positive \`tabindex\` values as they are error prone and can severely disrupt navigation experience for keyboard users. Use \`tabindex="0"\` to make an element focusable or \`tabindex="-1"\` to remove it from the tab sequence.`,
        attributeNode.location,
      )
    }
  }
}

export class HTMLNoPositiveTabIndexRule extends ParserRule {
  static ruleName = "html-no-positive-tab-index"

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "warning"
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new NoPositiveTabIndexVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}
